mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java

```diff
@@ -0,0 +1,151 @@
+    public static final int MINIMAL_MASK = 0x00;
+    public static final int MIME_DESCRIPTOR_MASK = 0x01;
+    public static final int HEADERS_MASK = 0x100;
+    public static final int FULL_CONTENT_MASK = 0x200;
+    public static final int BODY_CONTENT_MASK = 0x400;
+    public static final int MIME_HEADERS_MASK = 0x800;
+    public static final int MIME_CONTENT_MASK = 0x1000;
+
+    public static final FetchGroup MINIMAL = new FetchGroup(MINIMAL_MASK);
+    public static final FetchGroup HEADERS = new FetchGroup(HEADERS_MASK);
+    public static final FetchGroup FULL_CONTENT = new FetchGroup(FULL_CONTENT_MASK);
+    public static final FetchGroup BODY_CONTENT = new FetchGroup(BODY_CONTENT_MASK);
+
+    private final int content;
+    private final ImmutableSet<PartContentDescriptor> partContentDescriptors;
+
+    @VisibleForTesting
+    FetchGroup(int content) {
+        this(content, ImmutableSet.of());
+    }
+
+    @VisibleForTesting
+    FetchGroup(int content, ImmutableSet<PartContentDescriptor> partContentDescriptors) {
+        this.content = content;
+        this.partContentDescriptors = partContentDescriptors;
+    }
+
+    /**
+     * Contents to be fetched. Composed bitwise.
+     *
+     * @return bitwise description
+     * @see #MINIMAL_MASK
+     * @see #MIME_DESCRIPTOR_MASK
+     * @see #HEADERS_MASK
+     * @see #FULL_CONTENT_MASK
+     * @see #BODY_CONTENT_MASK
+     * @see #MIME_HEADERS_MASK
+     * @see #MIME_CONTENT_MASK
+     */
+    public int content() {
+        return content;
+    }
+
+    public FetchGroup or(int content) {
+         return new FetchGroup(this.content | content, partContentDescriptors);
+    }
+
+    /**
+     * Gets contents to be fetched for contained parts. For each part to be
+     * contained, only one descriptor should be contained.
+     *
+     * @return <code>Set</code> of {@link PartContentDescriptor}, or null if
+     *         there is no part content to be fetched
+     */
+    public Set<PartContentDescriptor> getPartContentDescriptors() {
+        return partContentDescriptors;
+    }
+
+    /**
+     * Adds content for the particular part.
+     * 
+     * @param path
+     *            <code>MimePath</code>, not null
+     * @param content
+     *            bitwise content constant
+     */
+    public FetchGroup addPartContent(MimePath path, int content) {
+        PartContentDescriptor newContent = partContentDescriptors.stream()
+            .filter(descriptor -> path.equals(descriptor.path()))
+            .findFirst()
+            .orElse(new PartContentDescriptor(path))
+            .or(content);
+
+        return new FetchGroup(this.content, ImmutableSet.<PartContentDescriptor>builder()
+            .addAll(partContentDescriptors.stream()
+                .filter(descriptor -> descriptor.path().equals(path))
+                .collect(Guavate.toImmutableSet()))
+            .add(newContent)
+            .build());
+    }
+
+    public boolean hasMask(int mask) {
```

Comment:
this method is crazy: how am I supposed to know what it does?
Do you expect to check the `content` bitfield contains at least the given mask?
Does it take any mask?
Only a power of two?
Do we check it contains at least one of the bits passed?
To be honest, I have bitfield but I can live with it. However, I don't think it's wise to expose public methods to do bitwise operations without a very good documentation + an extensive testsuite.
Given how costly it will be to implement, turning all this mess into a EnumSet would be much much faster.
I can even write it if you want to.

Reply:
To be really honest I want to do a full rewritte with an enumset, but I dont want to do it in this pr that is already way too big.

Can we live with bitewise descriptor in the short term, until I can find time to contribute that refactoring?